internal/client/client.go

```diff
@@ -175,126 +175,128 @@ func (c *CSAPI) SyncUntil(t *testing.T, since, key string, check func(gjson.Resu
-	}
-	return res
-}
-
-// MustDo is the same as Do but fails the test if the response is not 2xx
+// MustDo will do the HTTP request and fail the test if the response is not 2xx
 func (c *CSAPI) MustDo(t *testing.T, method string, paths []string, jsonBody interface{}) *http.Response {
 	t.Helper()
-	res, err := c.DoWithAuth(t, method, paths, jsonBody)
-	if err != nil {
-		t.Fatalf("CSAPI.MustDo %s %s error: %s", method, strings.Join(paths, "/"), err)
-	}
+	res := c.DoFunc(t, method, paths, WithJSONBody(t, jsonBody))
 	if res.StatusCode < 200 || res.StatusCode >= 300 {
 		t.Fatalf("CSAPI.MustDo %s %s returned HTTP %d", method, res.Request.URL.String(), res.StatusCode)
 	}
 	return res
 }
 
-func (c *CSAPI) MustDoRaw(t *testing.T, method string, paths []string, body []byte, contentType string, query url.Values) *http.Response {
-	t.Helper()
-	res, err := c.DoWithAuthRaw(t, method, paths, body, contentType, query)
-	if err != nil {
-		t.Fatalf("CSAPI.MustDoRaw %s %s error: %s", method, strings.Join(paths, "/"), err)
+// WithRawBody sets the HTTP request body to `body`
+func WithRawBody(body []byte) RequestOpt {
+	return func(req *http.Request) {
+		req.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+		// we need to manually set this because we don't set the body
+		// in http.NewRequest due to using functional options, and only in NewRequest
+		// does the stdlib set this for us.
+		req.ContentLength = int64(len(body))
 	}
-	if res.StatusCode < 200 || res.StatusCode >= 300 {
-		t.Fatalf("CSAPI.MustDoRaw %s %s returned HTTP %d", method, res.Request.URL.String(), res.StatusCode)
+}
+
+// WithContentType sets the HTTP request Content-Type header to `cType`
+func WithContentType(cType string) RequestOpt {
+	return func(req *http.Request) {
+		req.Header.Set("Content-Type", cType)
 	}
-	return res
 }
 
-// DoWithAuth a JSON request. The access token for this user will automatically be added.
-func (c *CSAPI) DoWithAuth(t *testing.T, method string, paths []string, jsonBody interface{}) (*http.Response, error) {
-	t.Helper()
-	var query url.Values
-	if c.AccessToken != "" {
-		query = url.Values{
-			"access_token": []string{c.AccessToken},
+// WithJSONBody sets the HTTP request body to the JSON serialised form of `obj`
+func WithJSONBody(t *testing.T, obj interface{}) RequestOpt {
+	return func(req *http.Request) {
+		t.Helper()
+		b, err := json.Marshal(obj)
+		if err != nil {
+			t.Fatalf("CSAPI.Do failed to marshal JSON body: %s", err)
 		}
+		WithRawBody(b)(req)
 	}
-	return c.Do(t, method, paths, jsonBody, query)
 }
 
-func (c *CSAPI) DoWithAuthRaw(t *testing.T, method string, paths []string, body []byte, contentType string, query url.Values) (*http.Response, error) {
-	t.Helper()
-	if query == nil {
-		query = url.Values{}
+// WithQueries sets the query parameters on the request. Does not modify access token as it
+// is set as an Authorization header.
+func WithQueries(q url.Values) RequestOpt {
```

Comment:
Shouldn't this be `WithQuery` or `WithQueryValues` instead of `WithQueries`? I think there's only one query, which is composed of multiple query keys and values.

Reply:
Well technically the term is query parameters. I don't want it to be too long so `WithQueryValues` is icky. I also don't want people to think they can only set 1 query parameter in the function which is why I went for `WithQueries`.

Reply:
Aesthetics. I'm all for clarity, but honestly it's a bikeshed. `WithQuery`, `WithQueries`, `WithQueryValues` are all clear enough, especially in the context of `DoFunc`. It's a heavily nested function as it is due to people calling `WithXXX` lots, so reducing the amount of useless characters is reasonable.